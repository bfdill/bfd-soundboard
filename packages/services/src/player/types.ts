import { ChildProcess } from 'child_process'

export type DefaultPlayer =
  | 'mplayer'
  | 'afplay'
  | 'mpg123'
  | 'mpg321'
  | 'play'
  | 'omxplayer'
  | 'aplay'
  | 'cmdmp3'
export type What = string | string[]
export type PlayerOptions = {
  player: any
  players?: DefaultPlayer | DefaultPlayer[] | string[]
}
export type PlayerArgs = { [playerName: string]: string[] }
export type PlayOptions = PlayerArgs & { stdio: 'ignore' }
export type Next = Function
export interface IPlayer {
  play: (what: What, options?: PlayOptions | Next, next?: Next) => ChildProcess
}
export type createPlayer = (opts?: PlayerOptions) => IPlayer
