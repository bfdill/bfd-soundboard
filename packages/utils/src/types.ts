export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : T[P] extends (string | number | boolean)
    ? T[P]
    : DeepPartial<T[P]>
}
