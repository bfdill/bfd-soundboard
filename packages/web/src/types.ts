import { Action } from 'redux'
import { ThunkAction } from 'redux-thunk'

export type FullState = {
  sounds: any
}

export type RequestGetSoundsAction = Action<'REQUEST/getSounds'>
export type SuccessGetSoundsAction = Action<'SUCCESS/getSounds'> & {
  payload: any
}
export type FailGetSoundsAction = Action<'FAILED/getSounds'> & { err: Error }
export type AnyAction =
  | RequestGetSoundsAction
  | SuccessGetSoundsAction
  | FailGetSoundsAction
export type GetSounds = () => ThunkAction<
  void,
  FullState,
  undefined,
  RequestGetSoundsAction | SuccessGetSoundsAction | FailGetSoundsAction
>
