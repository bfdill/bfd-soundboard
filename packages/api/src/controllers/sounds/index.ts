import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '@talkboy-9000/sounds'

import { IApplicationContext, ApplicationState } from '../../types'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds(logger)

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(SoundService.getInstance())

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
