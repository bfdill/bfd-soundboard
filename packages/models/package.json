{
  "name": "@talkboy-9000/models",
  "version": "0.0.1",
  "license": "Unlicense",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc -p ./tsconfig.json",
    "clean": "rimraf dist/"
  }
}
