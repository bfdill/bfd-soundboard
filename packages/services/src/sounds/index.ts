import { join } from 'path'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')
