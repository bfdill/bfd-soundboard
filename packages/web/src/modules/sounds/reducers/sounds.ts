import { Reducer } from 'redux'
import { AnyAction, GetSoundsState } from '../../../types'

const reducer: Reducer<GetSoundsState, AnyAction> = (
  state = { sounds: [] },
  action
) => {
  switch (action.type) {
    case 'SUCCESS/getSounds':
      return { sounds: action.payload }
  }

  return state
}

export default reducer
