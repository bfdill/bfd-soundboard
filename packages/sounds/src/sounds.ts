import * as sane from 'sane'
import * as winston from 'winston'
import { sync } from 'glob'
import { join, resolve, sep, basename } from 'path'
import { ISoundService, LOGGER_META } from './sounds.types'
import { Sound } from '@talkboy-9000/models'
import {
  Sounds as SoundConfig,
  ConfigurationService
} from '@talkboy-9000/configuration'

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly config: SoundConfig,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.config.PathToSounds, this.config.FileGlob)).forEach(
      this.addSound
    )
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  private createWatch = (): sane.Watcher => {
    return this.saneFunction(this.config.PathToSounds, {
      glob: this.config.FileGlob
    })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, this.config.PathToSounds.length) ===
      this.config.PathToSounds

    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: this.config.PathToSounds
      })

    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      ConfigurationService.getInstance().getSounds(),
      sane,
      winston.createLogger({ defaultMeta: LOGGER_META })
    )

    return this.instance
  }
}
