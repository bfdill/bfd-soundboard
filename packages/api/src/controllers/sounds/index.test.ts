import { Sound } from '@talkboy-9000/models'
import { ISoundService } from '@talkboy-9000/sounds'
import { getJestLogger } from '@talkboy-9000/winston-jest'
import { IApplicationContext } from '../../types'
import { SoundsController, getSoundsRouter } from '.'

describe('controllers -> sounds -> index', () => {
  const mockSound: Sound = {
    basename: 'mock basename',
    filename: 'mock filename',
    id: 'mock id'
  }

  test('has known exports', () => {
    expect(Object.keys(require('.'))).toMatchSnapshot()
  })

  test.each([null as any, [mockSound]])(
    'get matches snapshot',
    async (sounds?: Sound[] | null) => {
      const mockSoundService: ISoundService = {
        getSounds: () => sounds
      } as any
      const controller = new SoundsController(mockSoundService)
      const jestLogger = getJestLogger()
      const context: IApplicationContext = {
        logger: jestLogger.logger
      } as any

      await controller.get(context)

      expect(context.body).toMatchSnapshot()
      expect(context.status).toMatchSnapshot()
      jestLogger.callsMatchSnapshot()
    }
  )

  test('getSoundsRouter', () => {
    expect(getSoundsRouter()).toMatchSnapshot()
  })
})
