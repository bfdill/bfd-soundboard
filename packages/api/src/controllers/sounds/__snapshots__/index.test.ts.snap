// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers -> sounds -> index get matches snapshot 1`] = `
Object {
  "sounds": null,
}
`;

exports[`controllers -> sounds -> index get matches snapshot 2`] = `404`;

exports[`controllers -> sounds -> index get matches snapshot 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "get",
      "name": "SoundsController",
    },
    "level": "warn",
    "message": "no sounds from service",
    Symbol(level): "warn",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"SoundsController\\",\\"method\\":\\"get\\"},\\"message\\":\\"no sounds from service\\",\\"level\\":\\"warn\\"}",
  },
]
`;

exports[`controllers -> sounds -> index get matches snapshot 4`] = `
Object {
  "sounds": Array [
    Object {
      "basename": "mock basename",
      "filename": "mock filename",
      "id": "mock id",
    },
  ],
}
`;

exports[`controllers -> sounds -> index get matches snapshot 5`] = `200`;

exports[`controllers -> sounds -> index get matches snapshot 6`] = `
Array [
  Object {
    "controller": Object {
      "method": "get",
      "name": "SoundsController",
    },
    "level": "info",
    "message": Object {
      "soundsLength": 1,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"SoundsController\\",\\"method\\":\\"get\\"},\\"message\\":{\\"soundsLength\\":1},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> sounds -> index getSoundsRouter 1`] = `
Router {
  "methods": Array [
    "HEAD",
    "OPTIONS",
    "GET",
    "PUT",
    "PATCH",
    "POST",
    "DELETE",
  ],
  "opts": Object {},
  "params": Object {},
  "stack": Array [
    Layer {
      "methods": Array [
        "HEAD",
        "GET",
      ],
      "name": null,
      "opts": Object {
        "end": true,
        "ignoreCaptures": undefined,
        "name": null,
        "prefix": "",
        "sensitive": false,
        "strict": false,
      },
      "paramNames": Array [],
      "path": "/",
      "regexp": /\\^\\(\\?:\\\\/\\(\\?=\\$\\)\\)\\?\\$/i,
      "stack": Array [
        [Function],
      ],
    },
  ],
}
`;

exports[`controllers -> sounds -> index has known exports 1`] = `
Array [
  "SoundsController",
  "getSoundsRouter",
]
`;
